cps/src/main/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction.java

```diff
@@ -0,0 +1,242 @@
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.replay.Messages;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+/**
+ * Attached to a {@link Run} when it could be replayed with script edits.
+ */
+@SuppressWarnings("rawtypes") // on Run
+public class ReplayAction implements Action {
+
+    private final Run run;
+
+    private ReplayAction(Run run) {
+        this.run = run;
+    }
+
+    @Override public String getDisplayName() {
+        return "Replay";
+    }
+
+    @Override public String getIconFileName() {
+        return isEnabled() ? "clock.png" : null;
+    }
+
+    @Override public String getUrlName() {
+        return isEnabled() ? "replay" : null;
+    }
+
+    private @CheckForNull CpsFlowExecution getExecution() {
+        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) run).asFlowExecutionOwner();
+        if (owner == null) {
+            return null;
+        }
+        FlowExecution exec = owner.getOrNull();
+        return exec instanceof CpsFlowExecution ? (CpsFlowExecution) exec : null;
+    }
+
+    /* accessible to Jelly */ public boolean isEnabled() {
+        if (!run.hasPermission(REPLAY)) {
+            return false;
+        }
+        CpsFlowExecution exec = getExecution();
+        if (exec == null) {
+            return false;
+        }
+        if (exec.isSandbox()) {
+            return true;
+        } else {
+            // Whole-script approval mode. Can we submit an arbitrary script right here?
+            return Jenkins.getActiveInstance().hasPermission(Jenkins.RUN_SCRIPTS);
+        }
+    }
+
+    /* accessible to Jelly */ public String getOriginalScript() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getScript() : "???";
+    }
+
+    /* accessible to Jelly */ public Map<String,String> getOriginalLoadedScripts() {
+        CpsFlowExecution execution = getExecution();
+        return execution != null ? execution.getLoadedScripts() : /* ? */Collections.<String,String>emptyMap();
+    }
+
+    /* accessible to Jelly */ public Run getOwner() {
+        return run;
+    }
+
+    @RequirePOST
+    public void doRun(StaplerRequest req, StaplerResponse rsp) throws ServletException, IOException {
+        if (!isEnabled()) {
+            throw new AccessDeniedException("not allowed to replay"); // AccessDeniedException2 requires us to look up the specific Permission
+        }
+        JSONObject form = req.getSubmittedForm();
+        Map<String,String> otherScripts = new HashMap<String,String>();
+        for (String clazz : getOriginalLoadedScripts().keySet()) {
+            otherScripts.put(clazz, form.getString(clazz));
```

Comment:
Yes, `JSONException`. Conceivable in case you try to replay a _running_ build; in such a case I can just make it fall back to the original script.